lib/cookie-support.ts

```diff
@@ -0,0 +1,75 @@
+/**
+ * (C) Copyright IBM Corp. 2022.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import extend from 'extend';
+import { AxiosRequestConfig, AxiosResponse } from 'axios';
+import { CookieJar } from 'tough-cookie';
+import logger from './logger';
+
+export class CookieInterceptor {
+  private readonly cookieJar: CookieJar;
+
+  constructor(cookieJar: CookieJar | boolean) {
+    if (cookieJar) {
+      if (cookieJar === true) {
+        logger.debug('CookieInterceptor: creating new CookieJar');
+        this.cookieJar = new CookieJar();
+      } else {
+        logger.debug('CookieInterceptor: using supplied CookieJar');
+        this.cookieJar = cookieJar;
+      }
+    } else {
+      throw new Error('Must supply a cookie jar or true.');
+    }
+  }
+
+  public async requestInterceptor(config: AxiosRequestConfig) {
+    logger.debug('CookieInterceptor: intercepting request');
+    if (config && config.url) {
+      logger.debug(`CookieInterceptor: getting cookies for: ${config.url}`);
+      const cookieHeaderValue = await this.cookieJar.getCookieString(config.url);
+      if (cookieHeaderValue) {
+        logger.debug('CookieInterceptor: setting cookie header');
+        const cookieHeader = { cookie: cookieHeaderValue };
+        config.headers = extend(true, {}, config.headers, cookieHeader);
+      } else {
+        logger.debug(`CookieInterceptor: no cookies for: ${config.url}`);
+      }
+    } else {
+      logger.debug('CookieInterceptor: no request URL.');
+    }
+    return config;
+  }
+
+  public async responseInterceptor(response: AxiosResponse) {
+    logger.debug('CookieInterceptor: intercepting response.');
+    if (response && response.headers) {
+      logger.debug('CookieInterceptor: checking for set-cookie headers.');
+      const cookies: string[] = response.headers['set-cookie'];
+      if (cookies) {
+        logger.debug(`CookieInterceptor: setting cookies in jar for URL ${response.config.url}.`);
+        cookies.forEach(async (cookie: string) => {
+          await this.cookieJar.setCookie(cookie, response.config.url);
```

Comment:
There can be [some issues](https://stackoverflow.com/questions/37576685/using-async-await-with-a-foreach-loop?page=1&tab=scoredesc#tab-top) with calling `async` functions within a `forEach` method. Are you intending to call `setCookie` sequentially?

Reply:
>Are you intending to call `setCookie` sequentially?

My real intention was to ensure that all the cookie writes had completed prior to the interceptor resolving and returning a response so that all the cookies would be readable by subsequent requests that happen after a cookie response.
I honestly hadn't given a lot of thought to whether the requests were running in parallel or sequentially and I don't know too much about the behaviour of the async cookie stores and whether they safely handle parallel writes. So given that in the majority of cases there is likely only a single cookie anyway and cookies are small writing them sequentially seems probably the safest thing to do. I've made an attempt at that in ca9c3045.

Reply:
>Are you intending to call setCookie sequentially?

Good spot, thanks. My main intention was to ensure that the cookies were written before the interceptor returned the response, such that any requests happening after that response would be able to read the cookies.

I honestly hadn't given a lot of thought to whether writes were happening sequentially or in parallel and I don't know a lot about the safety of parallel writes to asynchronous cookie stores (of which I guess there are potentially many implementations). With that safety in mind and given that it is likely that the cookies are small and there will be only a single cookie in the majority of cases I suspect that there won't be a huge performance penalty from writing sequentially. So I've made an attempt at changing to that in 3bf097b.